homeassistant/loader.py

```diff
@@ -1419,18 +1419,24 @@ async def async_get_integrations(
             _resolve_integrations_from_root, hass, components, needed
         )
         for domain, future in needed.items():
-            int_or_exc = integrations.get(domain)
-            if not int_or_exc:
-                cache.pop(domain)
-                results[domain] = IntegrationNotFound(domain)
-            elif isinstance(int_or_exc, Exception):
-                cache.pop(domain)
+            if integration := integrations.get(domain):
+                results[domain] = cache[domain] = integration
+                future.set_result(integration)
+            else:
+                # We don't cache that it doesn't exist as configuration
+                # validation that relies on integrations being loaded
+                # would be unfixable. For example if a custom integration
+                # was temporarily removed.
+                # This allows restoring a missing integration to fix the
+                # validation error so the config validations checks do not
+                # block restarting.
+                del cache[domain]
                 exc = IntegrationNotFound(domain)
-                exc.__cause__ = int_or_exc
                 results[domain] = exc
-            else:
-                results[domain] = cache[domain] = int_or_exc
-            future.set_result(None)
+                # We don't use set_exception because
+                # we expect there will be cases where
+                # the a future exception is never retrieved
```

Comment:
where the future exception

Reply:
https://github.com/home-assistant/core/pull/139276